File: lib/core/utils/shop_time.dart
/// Shop-timezone helpers.
///
/// The shop runs on one configured timezone (`settings/general`, default
/// Asia/Manila UTC+8). Every calendar computation — "what day is it",
/// range boundaries, day grouping, display — must happen in that zone
/// rather than the device's. A phone set to another timezone otherwise
/// computes the wrong business day, and its `drawer_state` write is
/// rejected by the security rules: a hard sale failure, not a cosmetic bug.
///
/// Two representations, kept strictly apart:
///
/// * **instant** — a real point in time. What Firestore stores and what
///   `Timestamp.toDate()` returns. Never shifted before writing.
/// * **shop wall time** — an instant shifted by the shop offset so its
///   *fields* (`year`/`month`/`day`/`hour`) read shop-local. Produced by
///   [shopTimeOf]; always `isUtc` so field access is device-independent.
///   For day math and display ONLY — never write one to Firestore. Convert
///   back with [instantOf] before building a query bound or a `Timestamp`.
///
/// Only fixed-offset (no-DST) zones are supported, so plain offset
/// arithmetic is exact — see `shop_timezones.dart`.
library;

/// Asia/Manila — UTC+8, no DST. The system default everywhere.
const int kDefaultShopOffsetMinutes = 480;
const String kDefaultShopTimezoneId = 'Asia/Manila';

/// Ambient shop timezone, for code that cannot reach a Riverpod `ref`
/// (extension getters, formatting helpers). Set once at startup from the
/// cached value and again whenever `settings/general` changes — the same
/// pattern as `Intl.defaultLocale`. Riverpod code should prefer
/// `shopOffsetProvider` / `shopNowProvider` so tests can override it
/// without touching global state.
class ShopTimeConfig {
  ShopTimeConfig._();

  static int offsetMinutes = kDefaultShopOffsetMinutes;
  static String timezoneId = kDefaultShopTimezoneId;

  static void apply({required String timezoneId, required int offsetMinutes}) {
    ShopTimeConfig.timezoneId = timezoneId;
    ShopTimeConfig.offsetMinutes = offsetMinutes;
  }
}

/// Shop wall-clock view of [instant]. Its fields read shop-local. Do not persist.
DateTime shopTimeOf(DateTime instant, int offsetMinutes) =>
    instant.toUtc().add(Duration(minutes: offsetMinutes));

/// The real instant for a shop wall-clock time — inverse of [shopTimeOf].
DateTime instantOf(DateTime shopWallTime, int offsetMinutes) => DateTime.utc(
      shopWallTime.year,
      shopWallTime.month,
      shopWallTime.day,
      shopWallTime.hour,
      shopWallTime.minute,
      shopWallTime.second,
      shopWallTime.millisecond,
    ).subtract(Duration(minutes: offsetMinutes));

/// Builds a shop wall-clock value from calendar fields.
DateTime shopWall(
  int year, [
  int month = 1,
  int day = 1,
  int hour = 0,
  int minute = 0,
  int second = 0,
  int millisecond = 0,
]) =>
    DateTime.utc(year, month, day, hour, minute, second, millisecond);

/// yyyymmdd int for [instant] in shop time. Must equal the rules' `phDay()`
/// for the same instant — the `drawer_state` write depends on it.
int shopDayInt(DateTime instant, int offsetMinutes) {
  final w = shopTimeOf(instant, offsetMinutes);
  return w.year * 10000 + w.month * 100 + w.day;
}

/// Shop midnight (a wall value) of the day containing [instant].
DateTime shopDateOf(DateTime instant, int offsetMinutes) {
  final w = shopTimeOf(instant, offsetMinutes);
  return shopWall(w.year, w.month, w.day);
}

/// Ambient-offset conveniences for display and comparison code.
extension ShopTimeX on DateTime {
  /// Shop wall-clock view using the ambient offset. Display/day math only.
  DateTime get inShopTime => shopTimeOf(this, ShopTimeConfig.offsetMinutes);

  /// Treats this as a shop wall-clock value and returns the real instant.
  DateTime get asShopInstant => instantOf(this, ShopTimeConfig.offsetMinutes);
}

File: web_admin/src/domain/time/shopTime.ts
// Shop-timezone helpers — the TypeScript half of lib/core/utils/shop_time.dart.
//
// The shop runs on one configured timezone (settings/general, default
// Asia/Manila UTC+8). Every calendar computation — "what day is it", range
// boundaries, day grouping — uses that zone rather than the browser's, so a
// laptop set to another timezone still agrees with the mobile app and with
// the Firestore rules' phDay().
//
// Two representations, kept strictly apart:
//   * instant — a real point in time. What Firestore stores. Never shifted
//     before writing.
//   * shop wall time — an instant shifted by the shop offset so its UTC
//     getters read shop-local. For day math and display only; convert back
//     with instantOf() before using it as a query bound.
//
// Only fixed-offset (no-DST) zones are supported, so plain offset arithmetic
// is exact. Do NOT use date-fns' startOfDay/endOfDay on wall values — they
// read local getters and would double-shift.

export const DEFAULT_SHOP_OFFSET_MINUTES = 480;
export const DEFAULT_SHOP_TIMEZONE_ID = 'Asia/Manila';

export interface ShopTimezone {
  timezoneId: string;
  offsetMinutes: number;
}

const MINUTE_MS = 60_000;

let ambient: ShopTimezone = {
  timezoneId: DEFAULT_SHOP_TIMEZONE_ID,
  offsetMinutes: DEFAULT_SHOP_OFFSET_MINUTES,
};

/** Set from the settings doc at bootstrap and on every change. */
export function setAmbientShopTimezone(tz: ShopTimezone): void {
  ambient = tz;
}

export function getAmbientShopTimezone(): ShopTimezone {
  return ambient;
}

export function shopOffsetMinutes(): number {
  return ambient.offsetMinutes;
}

/** Shop wall-clock view of `instant`. Read it with getUTC*. Do not persist. */
export function shopTimeOf(instant: Date, offsetMinutes = shopOffsetMinutes()): Date {
  return new Date(instant.getTime() + offsetMinutes * MINUTE_MS);
}

/** The real instant for a shop wall-clock value — inverse of shopTimeOf. */
export function instantOf(wall: Date, offsetMinutes = shopOffsetMinutes()): Date {
  return new Date(wall.getTime() - offsetMinutes * MINUTE_MS);
}

/** Builds a shop wall-clock value from calendar fields (month is 1-based). */
export function shopWall(
  year: number,
  month: number,
  day: number,
  hour = 0,
  minute = 0,
  second = 0,
  ms = 0,
): Date {
  return new Date(Date.UTC(year, month - 1, day, hour, minute, second, ms));
}

/** yyyymmdd int in shop time — must equal the rules' phDay(). */
export function shopDayInt(instant: Date, offsetMinutes = shopOffsetMinutes()): number {
  const w = shopTimeOf(instant, offsetMinutes);
  return w.getUTCFullYear() * 10000 + (w.getUTCMonth() + 1) * 100 + w.getUTCDate();
}

/** Instant of shop midnight for the shop day containing `instant`. */
export function shopStartOfDay(instant: Date, offsetMinutes = shopOffsetMinutes()): Date {
  const w = shopTimeOf(instant, offsetMinutes);
  return instantOf(
    shopWall(w.getUTCFullYear(), w.getUTCMonth() + 1, w.getUTCDate()),
    offsetMinutes,
  );
}

/** Instant of the last millisecond of that shop day. */
export function shopEndOfDay(instant: Date, offsetMinutes = shopOffsetMinutes()): Date {
  const w = shopTimeOf(instant, offsetMinutes);
  return instantOf(
    shopWall(w.getUTCFullYear(), w.getUTCMonth() + 1, w.getUTCDate(), 23, 59, 59, 999),
    offsetMinutes,
  );
}

/** YYYYMMDD in shop time — the sale-counter key. */
export function shopDateKey(instant: Date, offsetMinutes = shopOffsetMinutes()): string {
  return `${shopDayInt(instant, offsetMinutes)}`;
}

/** yyyy-MM-dd in shop time — closing doc ids, pay-period dates. */
export function shopIsoDate(instant: Date, offsetMinutes = shopOffsetMinutes()): string {
  const w = shopTimeOf(instant, offsetMinutes);
  const y = w.getUTCFullYear();
  const m = `${w.getUTCMonth() + 1}`.padStart(2, '0');
  const d = `${w.getUTCDate()}`.padStart(2, '0');
  return `${y}-${m}-${d}`;
}

/** Formats an instant for display in the shop's zone. */
export function formatInShopZone(
  instant: Date,
  options: Intl.DateTimeFormatOptions,
  timezoneId = ambient.timezoneId,
): string {
  return new Intl.DateTimeFormat('en-PH', { ...options, timeZone: timezoneId }).format(instant);
}
